crates/transport/src/actors/gateway.rs

```diff
@@ -58,16 +68,96 @@ impl Default for GatewayConfig {
             events_queue_size: 100,
             shutdown_timeout: DEFAULT_SHUTDOWN_TIMEOUT,
             worker_status_via_gossipsub: false,
+            portal_logs_collector_lookup_time: Duration::from_secs(60),
         }
     }
 }
 
 pub struct GatewayTransport {
-    _task_manager: TaskManager,
-    query_handle: StreamClientHandle,
+    swarm: Swarm<Wrapped<GatewayBehaviour>>,
+    logs_rx: Receiver<QueryFinished>,
+    events_tx: Sender<GatewayEvent>,
+    logs_handle: StreamClientHandle,
+    selector: u64,
 }
 
 impl GatewayTransport {
+    pub async fn run(mut self, portal_logs_collector_lookup_time: Duration, cancel_token: CancellationToken) {
+        log::info!("Starting worker P2P transport");
+        let mut interval = time::interval(portal_logs_collector_lookup_time);
+        loop {
+            tokio::select! {
+                _ = interval.tick() => {
+                    self.swarm.behaviour_mut().update_portal_logs_listeners();
+                    log::debug!("listeners requested");
+                },
+                _ = cancel_token.cancelled() => break,
+                ev = self.swarm.select_next_some() => self.on_swarm_event(ev),
+                Some(log) = self.logs_rx.recv() => self.publish_portal_logs(log).await,
+            }
+        }
+        log::info!("Shutting down worker P2P transport");
+    }
+
+    fn on_swarm_event(&mut self, ev: SwarmEvent<GatewayEvent>) {
+        log::trace!("Swarm event: {ev:?}");
+        record_event(&ev);
+        if let SwarmEvent::Behaviour(ev) = ev {
+            self.events_tx.send_lossy(ev)
+        }
+    }
+
+    async fn publish_portal_logs(&mut self, log: QueryFinished) {
+        log::debug!("Sending log: {log:?}");
+        let log_size = log.encoded_len() as u64;
+        if log_size > MAX_QUERY_MSG_SIZE {
```

Comment:
I think there is no reason for this check. I can see that other actors do it, but I was considering removing it. It's especially painful since you're not forwarding this error back to the caller.